File: codes/go/chapter_tree/binary_search_tree.go
// File: binary_search_tree.go
// Created Time: 2022-11-26
// Author: Reanon (793584285@qq.com)

package chapter_tree

import (
	"sort"

	. "github.com/krahets/hello-algo/pkg"
)

type binarySearchTree struct {
	root *TreeNode
}

func newBinarySearchTree(nums []int) *binarySearchTree {
	// 排序数组
	sort.Ints(nums)
	// 构建二叉搜索树
	bst := &binarySearchTree{}
	bst.root = bst.buildTree(nums, 0, len(nums)-1)
	return bst
}

/* 构建二叉搜索树 */
func (bst *binarySearchTree) buildTree(nums []int, left, right int) *TreeNode {
	if left > right {
		return nil
	}
	// 将数组中间节点作为根节点
	middle := left + (right-left)>>1
	root := NewTreeNode(nums[middle])
	// 递归构建左子树和右子树
	root.Left = bst.buildTree(nums, left, middle-1)
	root.Right = bst.buildTree(nums, middle+1, right)
	return root
}

/* 获取根节点 */
func (bst *binarySearchTree) getRoot() *TreeNode {
	return bst.root
}

/* 获取中序遍历的下一个节点（仅适用于 root 有左子节点的情况） */
func (bst *binarySearchTree) getInOrderNext(node *TreeNode) *TreeNode {
	if node == nil {
		return node
	}
	// 循环访问左子节点，直到叶节点时为最小节点，跳出
	for node.Left != nil {
		node = node.Left
	}
	return node
}

/* 查找节点 */
func (bst *binarySearchTree) search(num int) *TreeNode {
	node := bst.root
	// 循环查找，越过叶节点后跳出
	for node != nil {
		if node.Val < num {
			// 目标节点在 cur 的右子树中
			node = node.Right
		} else if node.Val > num {
			// 目标节点在 cur 的左子树中
			node = node.Left
		} else {
			// 找到目标节点，跳出循环
			break
		}
	}
	// 返回目标节点
	return node
}

/* 插入节点 */
func (bst *binarySearchTree) insert(num int) *TreeNode {
	cur := bst.root
	// 若树为空，直接提前返回
	if cur == nil {
		return nil
	}
	// 待插入节点之前的节点位置
	var pre *TreeNode = nil
	// 循环查找，越过叶节点后跳出
	for cur != nil {
		if cur.Val == num {
			return nil
		}
		pre = cur
		if cur.Val < num {
			cur = cur.Right
		} else {
			cur = cur.Left
		}
	}
	// 插入节点
	node := NewTreeNode(num)
	if pre.Val < num {
		pre.Right = node
	} else {
		pre.Left = node
	}
	return cur
}

/* 删除节点 */
func (bst *binarySearchTree) remove(num int) *TreeNode {
	cur := bst.root
	// 若树为空，直接提前返回
	if cur == nil {
		return nil
	}
	// 待删除节点之前的节点位置
	var pre *TreeNode = nil
	// 循环查找，越过叶节点后跳出
	for cur != nil {
		if cur.Val == num {
			break
		}
		pre = cur
		if cur.Val < num {
			// 待删除节点在右子树中
			cur = cur.Right
		} else {
			// 待删除节点在左子树中
			cur = cur.Left
		}
	}
	// 若无待删除节点，则直接返回
	if cur == nil {
		return nil
	}
	// 子节点数为 0 或 1
	if cur.Left == nil || cur.Right == nil {
		var child *TreeNode = nil
		// 取出待删除节点的子节点
		if cur.Left != nil {
			child = cur.Left
		} else {
			child = cur.Right
		}
		// 将子节点替换为待删除节点
		if pre.Left == cur {
			pre.Left = child
		} else {
			pre.Right = child
		}
		// 子节点数为 2
	} else {
		// 获取中序遍历中待删除节点 cur 的下一个节点
		next := bst.getInOrderNext(cur)
		temp := next.Val
		// 递归删除节点 next
		bst.remove(next.Val)
		// 将 next 的值复制给 cur
		cur.Val = temp
	}
	return cur
}

/* 打印二叉搜索树 */
func (bst *binarySearchTree) print() {
	PrintTree(bst.root)
}

File: codes/python/chapter_tree/binary_search_tree.py
"""
File: binary_search_tree.py
Created Time: 2022-12-20
Author: a16su (lpluls001@gmail.com)
"""

import sys, os.path as osp
sys.path.append(osp.dirname(osp.dirname(osp.abspath(__file__))))
from modules import *


class BinarySearchTree:
    """ 二叉搜索树 """
    def __init__(self, nums: list[int]) -> None:
        """ 构造方法 """
        nums.sort()
        self.__root = self.build_tree(nums, 0, len(nums) - 1)

    def build_tree(self, nums: list[int], start_index: int, end_index: int) -> TreeNode | None:
        """ 构建二叉搜索树 """
        if start_index > end_index:
            return None

        # 将数组中间节点作为根节点
        mid: int = (start_index + end_index) // 2
        root = TreeNode(nums[mid])
        # 递归建立左子树和右子树
        root.left = self.build_tree(nums=nums, start_index=start_index, end_index=mid - 1)
        root.right = self.build_tree(nums=nums, start_index=mid + 1, end_index=end_index)
        return root

    @property
    def root(self) -> TreeNode | None:
        return self.__root

    def search(self, num: int) -> TreeNode | None:
        """ 查找节点 """
        cur: TreeNode | None = self.__root
        # 循环查找，越过叶节点后跳出
        while cur is not None:
            # 目标节点在 cur 的右子树中
            if cur.val < num:
                cur = cur.right
            # 目标节点在 cur 的左子树中
            elif cur.val > num:
                cur = cur.left
            # 找到目标节点，跳出循环
            else:
                break
        return cur

    def insert(self, num: int) -> TreeNode | None:
        """ 插入节点 """
        # 若树为空，直接提前返回
        if self.__root is None:
            return None
        
        # 循环查找，越过叶节点后跳出
        cur, pre = self.__root, None
        while cur is not None:
            # 找到重复节点，直接返回
            if cur.val == num:
                return None
            pre = cur
            # 插入位置在 cur 的右子树中
            if cur.val < num:
                cur = cur.right
            # 插入位置在 cur 的左子树中
            else:
                cur = cur.left

        # 插入节点 val
        node = TreeNode(num)
        if pre.val < num:
            pre.right = node
        else:
            pre.left = node
        return node

    def remove(self, num: int) -> TreeNode | None:
        """ 删除节点 """
        # 若树为空，直接提前返回
        if self.__root is None:
            return None

        # 循环查找，越过叶节点后跳出
        cur, pre = self.__root, None
        while cur is not None:
            # 找到待删除节点，跳出循环
            if cur.val == num:
                break
            pre = cur
            if cur.val < num:  # 待删除节点在 cur 的右子树中
                cur = cur.right
            else:  # 待删除节点在 cur 的左子树中
                cur = cur.left
        # 若无待删除节点，则直接返回
        if cur is None:
            return None

        # 子节点数量 = 0 or 1
        if cur.left is None or cur.right is None:
            # 当子节点数量 = 0 / 1 时， child = null / 该子节点
            child = cur.left or cur.right
            # 删除节点 cur
            if pre.left == cur:
                pre.left = child
            else:
                pre.right = child
        # 子节点数量 = 2
        else:
            # 获取中序遍历中 cur 的下一个节点
            nex: TreeNode = self.get_inorder_next(cur.right)
            tmp: int = nex.val
            # 递归删除节点 nex
            self.remove(nex.val)
            # 将 nex 的值复制给 cur
            cur.val = tmp
        return cur

    def get_inorder_next(self, root: TreeNode | None) -> TreeNode | None:
        """ 获取中序遍历中的下一个节点（仅适用于 root 有左子节点的情况） """
        if root is None:
            return root
        # 循环访问左子节点，直到叶节点时为最小节点，跳出
        while root.left is not None:
            root = root.left
        return root


""" Driver Code """
if __name__ == "__main__":
    # 初始化二叉搜索树
    nums = list(range(1, 16)) # [1, 2, ..., 15]
    bst = BinarySearchTree(nums=nums)
    print("\n初始化的二叉树为\n")
    print_tree(bst.root)

    # 查找节点
    node = bst.search(7)
    print("\n查找到的节点对象为: {}，节点值 = {}".format(node, node.val))

    # 插入节点
    node = bst.insert(16)
    print("\n插入节点 16 后，二叉树为\n")
    print_tree(bst.root)

    # 删除节点
    bst.remove(1)
    print("\n删除节点 1 后，二叉树为\n")
    print_tree(bst.root)

    bst.remove(2)
    print("\n删除节点 2 后，二叉树为\n")
    print_tree(bst.root)

    bst.remove(4)
    print("\n删除节点 4 后，二叉树为\n")
    print_tree(bst.root)
